### ISSUE TITLE
Extremely slow model imports

### ISSUE THREAD

**Issue by ofek** [pydantic/pydantic#9908]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

The following code is extremely slow to import:

```
❯ python -m timeit -n 1 -r 1 "from deva.env.models import EnvironmentStatus"
1 loop, best of 1: 165 msec per loop
```

(every `__init__.py` along the way is empty)

### Example Code

```Python
from typing import Literal

from pydantic import BaseModel, ConfigDict


class EnvironmentStatus(BaseModel):
    model_config = ConfigDict(strict=True, frozen=True)

    state: Literal['active', 'inactive', 'starting', 'stopping']
    info: str = ''
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: C:\Users\ofek\AppData\Local\Programs\Python\Python311\Lib\site-packages\pydantic
               python version: 3.11.7 (tags/v3.11.7:fa7a6f2, Dec  4 2023, 19:24:49) [MSC v.1937 64 bit (AMD64)]
                     platform: Windows-10-10.0.22631-SP0
             related packages: email-validator-2.1.0.post1 typing_extensions-4.9.0
                       commit: unknown
```

**Comment by ofek** [comment]
Unfortunately, we won't be able to use this in our internal tooling at work (and personally I couldn't introduce this in Hatch as I had planned on).

**Comment by mikedavidson-evertz** [comment]
I've created a issue for this here: https://github.com/pydantic/pydantic/issues/9905 - Just posting for visibility

**Comment by sydney-runkle** [comment]
@ofek,

Thanks for reporting this. I'm working on speeding up core schema build time (which will speed up import time) as well. Do you have a time for such an import that you'd find fast enough such that you could use `pydantic` for your tooling?

**Comment by ofek** [comment]
Oh that's great news, thanks! I tried `msgspec` on the same code and importing takes around 35-40 ms so I would expect that but 50-60 ms would be tolerable.

**Comment by ofek** [comment]
Out of curiosity, how can this be "fixed"? Just wondering what the plan is technically 🙂

**Comment by ofek** [comment]
There is no difference for me (also I'm on Windows)

**Comment by ofek** [comment]
![image](https://github.com/user-attachments/assets/5868808b-7847-44d9-b0e1-cf8910178005)

**Comment by sydney-runkle** [comment]
@ofek,

Huh, that's odd. We've been seeing improvements in our benchmarks for import time as well as schema build time...

**Comment by sydney-runkle** [comment]
And what are the results like when you're on v2.8.2?

**Comment by ofek** [comment]
![image](https://github.com/user-attachments/assets/51ef9c23-2714-4901-a8b2-d53005f42430)

**Comment by sneakers-the-rat** [comment]
if it's helpful to see different kinds of models that have slow imports, here's an example of a set of models that also have extremely slow imports (with version in `main` - `2.9.0a1@ddc8f58`) that i would expect to be very fast: https://github.com/linkml/linkml/blob/d9019caaf1d0bb60d2cd145e1441dec7d59ae1f9/linkml/generators/common/build.py

Each of these models only has a single field, which is a relatively complex dataclass.

```
$ python -m timeit -n 1 -r 1 "from linkml.generators.common import build"
1 loop, best of 1: 2.59 sec per loop
```

it seems like the schema generation recurses without memoization? while each of those dataclasses has many fields, many of them have the same type annotation, and even are referring to the same submodels, but they seem to be being build fresh every time. might make sense to use a cache there where generating schema for the same objects (maybe dataclasses are a special case because of the potential for deep recursion?) only happens once.

Only offering as an additional example if it's helpful, can split off into separate issue if thats better.

**Comment by sydney-runkle** [comment]
@sneakers-the-rat,

Indeed, we could benefit from a greater degree of memoization / caching. We cache model schemas, but right now our config structure doesn't support lots of type schema caching. We're continuing to work on this 👍.

Thanks for the example! That'll be helpful for our benchmarking!

**Comment by ofek** [comment]
This has only gotten worse on the latest versions. Here is Python 3.12.1:

```python
from typing import Literal

from pydantic import BaseModel, ConfigDict


class EnvironmentStatus(BaseModel):
    model_config = ConfigDict(strict=True, frozen=True)

    state: Literal['active', 'inactive', 'starting', 'stopping']
    info: str = ''
```

![image](https://github.com/user-attachments/assets/d52317d8-b079-4401-9ec3-27e149a4f63d)

**Comment by ofek** [comment]
The times don't even change when I reduce it to the following:

```python
from pydantic import BaseModel

class EnvironmentStatus(BaseModel):
    info: str = ''
```

**Comment by sydney-runkle** [comment]
Hmm, thanks for reporting. We reported a 35% improvement on model import times based on the benchmarks in this PR: https://github.com/pydantic/pydantic/pull/10009.

Happy to continue working on this - thanks for reporting your recent stats.

**Comment by ofek** [comment]
Can you reproduce locally using the snippet I provided?

**Comment by sydney-runkle** [comment]
I wasn't able to repro the same issues... perhaps you could do some profiling on your imports?

**Comment by ofek** [comment]
Could you show the same in your terminal? Create a directory and then enter the directory and then create a file named `test.py` with the following contents:

```python
from pydantic import BaseModel

class Foo(BaseModel):
    field: str = ''
```

```
❯ python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 174 msec per loop
❯ python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 182 msec per loop
❯ python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 177 msec per loop
```

You can also reproduce this in Docker by creating the following file:

```dockerfile
FROM python:3.12

COPY test.py test.py
RUN pip install pydantic==2.9.2
```

Then build it and run it:

```
docker build . --tag pydantic-test
docker run --rm -it pydantic-test bash
```

```
root@c0d4529233b1:/# python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 125 msec per loop
root@c0d4529233b1:/# python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 119 msec per loop
root@c0d4529233b1:/# python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 118 msec per loop
root@c0d4529233b1:/# pip install -qqq pydantic==2.8.2
root@c0d4529233b1:/# python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 110 msec per loop
root@c0d4529233b1:/# python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 102 msec per loop
root@c0d4529233b1:/# python -m timeit -n 1 -r 1 "from test import Foo"
1 loop, best of 1: 104 msec per loop
```

**Comment by ofek** [comment]
Were you able to run the example?

**Comment by sneakers-the-rat** [comment]
if it helps put a bound on what we can shoot for, profiling with `python -X importtime test_import_2.py` (python 3.12.6, pydantic 2.9.2) which imports a second module that defines the model above gets me this:

<img width="1634" alt="Screenshot 2024-10-02 at 7 20 32 PM" src="https://github.com/user-attachments/assets/aacae841-8a8a-4504-8eca-7bffc53fa216">

so if we subtract the import time of all the builtins to get the pydantic-specific import time (importing all the builtins is part of pydantic's import time, but you get what i mean) i get

|  | time (ms) |
| --- | --- |
|  total | 140 |
| importlib.metadata | 20 |
| inspect | 9 | 
| typing | 4 |
| pathlib | 4 |
| others | ~5 |
| **pydantic** | **98** |

That agrees roughly with what happens when i run timeit on just importing pydantic vs importing it and defining a model (which doesn't have the overhead of importtime's profiling):
* ~130ms with model definition
* ~60ms without model definition

Since in prior posts Ofek is saying he would want something like 50-60ms would be tolerable, it seems like there is some perf that would need to be gained in both. I'm sure y'all have profiled the heck out of the model construction here and are hunting for perf, i just wanted to say that it looks like we might also need to look at pruning the import tree a bit too. the `plugin._loader` that loads `importlib.metadata` is one obvious target for a quick exit if someone doesn't have any plugins defined (though it may be needed elsewhere, i don't know). `annotated_types` also contributes ~10ms which seems like a lot for what it is.

hopefully not annoying, i have just been profiling lately too and wanna help if i can

**Comment by sydney-runkle** [comment]
> hopefully not annoying, i have just been profiling lately too and wanna help if i can

This is great, thanks so much for reporting your findings!!

@ofek,

Apologies for the delay. We're sprinting for the next two weeks on v2.10 features, then we'll pivot back to a performance focus for the rest of the year. Would love to collaborate with you folks re getting these import times down for future minor releases!

**Comment by antont** [comment]
One test case would be the pydantic based atproto implementation where people have reported slow import times there https://github.com/MarshalX/atproto/issues/453#issuecomment-2494626029

**Comment by ofek** [comment]
@sydney-runkle Did you happen to run the example I posted yet in order to reproduce?

**Comment by sydney-runkle** [comment]
Hey @ofek,

Thanks for following up. Not yet - off for the holidays.

Building a big Q1 plan early next week. We'll incorporate perf improvements into that and circle back :).

**Comment by Viicos** [comment]
2.11 (final released planned by the end of March) brings significant performance improvements related to constructing models, in particular core schema building [^1].

An alpha release (2.11.0a1) was released today and can be installed, so that you can try it out. Do not hesitate to provide feedback if this improved (or did not) performance on your side.

Note that memory consumption haven’t been really optimized yet in this alpha release. We have plans to optimize it before the final release.

---

@ofek, with your provided snippet, I am down to ~44ms, while I have ~95ms on 2.10. However, I assume you will not be defining a single model. Even if we managed to improve model build performance even more, this isn't going to scale well for a CLI application.

One alternative path would be to set [`defer_build`](https://docs.pydantic.dev/dev/api/config/#pydantic.config.ConfigDict.defer_build) on your model configuration, and only have them built depending on the CLI command used (this can be done explicitly by calling [`model_rebuild()`](https://docs.pydantic.dev/dev/api/base_model/#pydantic.BaseModel.model_rebuild), but trying to instantiate a model for the first time will also trigger the build).

With the configuration set, it gets down to around ~35ms. For reference, a bare stdlib dataclass is around ~25ms.

[^1]: Pydantic defines a core schema whenever a Model is defined, read more [here](https://docs.pydantic.dev/latest/internals/architecture/)

**Comment by vjeranc** [comment]
@Viicos is it possible that `defer_build=True` hides some failures, which will then show up when program runs instead of startup?

**Comment by Viicos** [comment]
> hides some failures

What failures are you referring to?

**Comment by vjeranc** [comment]
> What failures are you referring to?

If schema is incorrectly specified.

**Comment by Viicos** [comment]
What schema are you referring to? If your question is unrelated to this issue, please [open a discussion](https://github.com/pydantic/pydantic/discussions/new/choose), as comments may be notifying users subscribed to this thread.

**Comment by vjeranc** [comment]
I'll make myself extremely clear summarizing the thread.

1. Imports are very slow due to a lot of classes or data types annotated by pydantic functions.

2. On my machine around 1000 invocations of create_data class take 4 seconds.

3. Your suggestion was to defer build to runtime and not import time. Meaning that when code invokes validation or parsing facilities it will build them then, not before.

4. Given that build might depend on configuration or other stuff, there's a chance that what was before error at import will now become error at runtime (if not tested)

5. My question aims at previous point and I'm wondering if that's indeed the case.

Fixing a problem by potentially introducing another problem is something that IMO isn't considered a fix.

**Comment by Viicos** [comment]
> 4\. Given that build might depend on configuration or other stuff, there's a chance that what was before error at import will now become error at runtime (if not tested)
> Fixing a problem by potentially introducing another problem is something that IMO isn't considered a fix.

Yes, but this is by definition. But not having tests is a bigger problem anyway.

On top of the build time improvements in 2.11, I suggested `defer_build` specifically for CLI applications, where depending on the command invoked, you only need a subset of your models/type hints to be built. Having tests for each CLI command implementation would cover any schema build errors.

**Comment by ofek** [comment]
@Viicos I feel like I'm going insane by getting different results locally compared to you and @sydney-runkle so I temporarily opened a [PR](https://github.com/pydantic/pydantic/pull/11527) that adds a simple job matrix to illustrate what I'm seeing. [Here](https://github.com/pydantic/pydantic/actions/runs/13691935019/job/38286574171?pr=11527) are the results, look for job names in the form `Pydantic X.Y.Z, Python X.Y, PLATFORM`.

Based on that, my guess is that both of your timings are taken on macOS because that appears to be faster whereas Linux and Windows match what I've been testing locally and in Docker.

Here are the results for Python 3.13 with timings in milliseconds:

| | | | |
| --- | --- | --- | --- |
| | Linux | Windows | macOS |
| 2.8.2 | 103 | 130 | 57 |
| 2.9.2 | 116 | 139 | 80 |
| 2.10.6 | 110 | 140 | 60 |
| 2.11.0a2 | 86 | 110 | 50 |

Which platforms are you testing on to come up with the timings you mentioned? An important thing to note from this is that what I said previously is 100% reproducible: 2.9 timings got worse across the board despite what was claimed. Thus, I think it's incredibly important to know how you both are testing these timings and seeing different results.

**Comment by Viicos** [comment]
Seems like I get different results from my initial experiment, but it also turns out that for such a simple model, the `from pydantic import BaseModel` import time is taking most of the time. On my machine, I get an average import time of 40ms, which is acceptable (although might be considered too much for CLI apps).

Looking at the import profile:

![Image](https://github.com/user-attachments/assets/cad2e372-ff10-4bcc-a58f-8170b97f9c24)

There's not much we can really improve. Pydantic depends on many stdlib modules (`re`, `pathlib`, `typing`/`typing-extensions`, `zoneinfo`, `fractions`, `ipaddress`, `uuid`) to check type hints against the related types.

What would be an acceptable limit in terms of delay for a CLI app in Python for you? Pydantic models are slower to build, but validation is faster compared to pure Python validation that could be peformed by other CLI frameworks such as Click. It might be that in the end, using Pydantic will be faster, but the opposite might be true as well.

**Comment by Viicos** [comment]
Closing as per what was said above.
